### DISCUSSION TITLE
How to use signal to pause and resume workflow ?

### DISCUSSION THREAD

**Discussion by lantian789** [uber/cadence#4503] [Q&A]
There is some sample code snippet of how to use signal in some relevant documentation:

1. https://cadenceworkflow.io/docs/java-client/signals/#implement-signal-handler-in-workflow
2. https://pkg.go.dev/go.uber.org/cadence@v0.18.2/workflow#hdr-Signals

It 's still very hard for me to know how to use this great feature to stop and resume the workflow, I really appreciate if someone could provide a complete demo code to help me master this wonderful technique. Thank you so much !

**Comment by longquanzheng** [comment]
Key idea:
1. Use a signal to modify a status variable to indicate whether or not it's paused 
2. Use `workflow.Go` to run a different goroutine(don't use native go to start, which would cause non-deterministic issues)
2. Use `workflow.Await` to wait for the condition before executing any step you want to check

Sample code in Golang client:
```golang
func samplePauseResumeWorkflow(ctx workflow.Context) error {
                // Use a signal to modify a status variable to indicate whether or not it's paused 
		s := workflow.NewSelector(ctx)
		paused := false
		s.AddReceive(workflow.GetSignalChannel(ctx, "operation"), func(c workflow.Channel, ok bool) {
			if ok{
				c.Receive(ctx, &paused)
			}
		})
		
		// Use `workflow.Go` to run a different goroutine(don't use native go to start, which would cause non-deterministic issues)
		workflow.Go(ctx, func(ctx workflow.Context) {
			for {
				s.Select(ctx)
			}	
		})
		
                // Use `workflow.Await` to wait for the condition before executing any step you want to check
		workflow.Await(ctx, func() bool {
			return paused != true
		})
		workflow.ExecuteActivity(ctx, activityA)
		
		workflow.Await(ctx, func() bool {
			return paused != true
		})
		workflow.ExecuteActivity(ctx, activityB)
		....
                 ....
}
```

**Reply by lantian789** [reply]
How to send signal to the workflow ?

**Reply by longquanzheng** [reply]
Use this api for CLI https://github.com/uber-go/cadence-client/blob/master/client/client.go#L141


https://cadenceworkflow.io/docs/cli/

**Reply by lantian789** [reply]
err = workflowClient.SignalWorkflow(context.Background(), sampleWorkflow.ID, sampleWorkflow.RunID, "operation", true)， Is this the right way to pass the true value to the workflow ?

**Reply by lantian789** [reply]
what is the purpose of doing "Use `workflow.Go` to run a different goroutine(don't use native go to start, which would cause non-deterministic issues)" ?

**Reply by longquanzheng** [reply]
That’s the similar idea of go func () … to start a gorotine to run something concurrently without blocking the next line of code. 

Without that, we can’t have the next line running and the code the in the forloop forever.

**Reply by longquanzheng** [reply]
Yeah that’s the right way to send signal

**Reply by lantian789** [reply]
I thought Await function will block the execution of the code until paused is set to false.

**Reply by Groxx** [reply]
await will block the calling goroutine, yes.  that's why the select is happening in another goroutine - that isn't blocked by await, so it can react to the signal.
